apps/web/src/components/BlocksEditor/Editor/state/promptlToLexical/fromAstToBlocks.ts

```diff
@@ -331,17 +358,33 @@ function processBlockNode({
       indent: 0,
       type: BLOCK_EDITOR_TYPE.STEP,
       errors: blockErrors?.length > 0 ? blockErrors : undefined,
-      // @ts-expect-error - Improve Typescript types for children
-      children:
-        children.length > 0
-          ? processNodes({ nodes: node.children, prompt, errors })
-          : [createEmptyParagraph({ content: '' })],
+      children: processedChildren,
     } satisfies StepBlock
   }
 
   if (isMessageBlock(node)) {
     const message = node as ElementTag
     const children = message.children ?? []
+    let processedChildren =
+      children.length > 0
+        ? (processNodes({ nodes: children, prompt, errors }) as (
+            | ParagraphBlock
+            | CodeBlock
+          )[])
+        : [createEmptyParagraph({ content: '' })]
+
+    // Normalize indentation added during pretty-print of previous roundtrips
+    processedChildren = normalizeIndentationInMessageChildren(processedChildren)
+
+    // Trim leading/trailing empty or whitespace-only paragraphs inside messages
+    processedChildren = trimEmptyMessageChildren(processedChildren)
+
+    // Collapse multiple consecutive empty paragraphs to a single one
+    processedChildren = collapseConsecutiveEmptyParagraphs(processedChildren)
+    if (processedChildren.length === 0) {
+      processedChildren = [createEmptyParagraph({ content: '' })]
```

Comment:
All of this is repeated twice, or looks super similar. Maybe do a helper method for this whitespace treatment